source/extensions/transport_sockets/tcp_stats/tcp_stats.cc

```diff
@@ -79,7 +79,7 @@ absl::optional<struct tcp_info> TcpStatsSocket::querySocketInfo() {
   memset(&info, 0, sizeof(info));
   socklen_t optlen = sizeof(info);
   const auto result = callbacks_->ioHandle().getOption(IPPROTO_TCP, TCP_INFO, &info, &optlen);
-  if ((result.return_value_ != 0) || (optlen < sizeof(info))) {
+  if (result.return_value_ != 0) {
```

Comment:
This change would silently not count the unpopulated fields.

What kernel version did this fail on? I'm guessing it's quite old. I think if you want to handle older kernels, you should figure out what is and is not valid in the tcp_info, and not create the stats for the fields that aren't supported, instead of publishing zeros for them with no indication that the values aren't being set correctly.

/wait

Reply:
4.15.0-162-generic.

If I understand it right, this code assumes the kernel where Envoy is built must match the kernel where it runs?

Reply:
Oh, I see what went wrong, I think. I looked at the version of `linux/tcp.h` in the build image and didn't think about other build environments.

The current assumption is that the running kernel version be >= kernel headers built against. I think we can substantially relax that by computing the offset of the last field we use in the struct, and setting optlen equal to that. Then we'll only require a kernel that provides the exact fields we're using. The kernel ABI guarantees that fields can be added, but no old fields will change size or be removed.

Reply:
We may not need to adjust the length, I think it may just be ok that unpopulated fields are 0. Some stats may be wrong on kernels which do not support all options but it should not be a problem otherwise.

Reply:
I personally like this approach better. The approach with offset of the last field we use is brittle. Someone can use a different field and have dig for why it is always 0. Functionally this approach and #22451 are the same but this one seems to be less error prone.

Reply:
@yanavlasov raises good points. I'm ok with keeping this approach.